pkg/jsonschema/unique_constraint.go

```diff
@@ -0,0 +1,229 @@
+			InstanceLocation: s.ERD.Name,
+			KeywordLocation:  "unique",
+			Message:          ErrUniqueConstraintViolation.Error(),
+		}
+	}
+
+	return nil
+}
+
+// UniqueConstraintCompiler is the compiler struct for the unique constraint JSON schema extension
+type UniqueConstraintCompiler struct {
+	ERD        *models.ExtensionResourceDefinition
+	ResourceID *string
+	ctx        context.Context
+	db         boil.ContextExecutor
+}
+
+// UniqueConstraintCompiler implements jsonschema.ExtCompiler
+var _ jsonschema.ExtCompiler = (*UniqueConstraintCompiler)(nil)
+
+// Compile compiles the unique constraint JSON schema extension
+func (uc *UniqueConstraintCompiler) Compile(
+	_ jsonschema.CompilerContext, m map[string]interface{},
+) (jsonschema.ExtSchema, error) {
+	unique, ok := m["unique"]
+	if !ok {
+		// If "unique" is not in the map, skip processing
+		return nil, nil
+	}
+
+	uniqueFields, err := assertStringSlice(unique)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(uniqueFields) == 0 {
+		// unique property is not provided, skip
+		return nil, nil
+	}
+
+	requiredFields, err := assertStringSlice(m["required"])
+	if err != nil {
+		return nil, err
+	}
+
+	requiredMap := make(map[string]bool, len(requiredFields))
+	for _, f := range requiredFields {
+		requiredMap[f] = true
+	}
+
+	propertiesMap, ok := m["properties"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf(
+			`%w: cannot apply unique constraint when "properties" is not provided or invalid`,
+			ErrInvalidUniqueProperty,
+		)
+	}
+
+	return uc.compileUniqueConstraint(uniqueFields, requiredMap, propertiesMap)
+}
+
+func (uc *UniqueConstraintCompiler) compileUniqueConstraint(
+	uniqueFields []string, requiredMap map[string]bool, propertiesMap map[string]interface{},
+) (jsonschema.ExtSchema, error) {
+	// map fieldName => fieldType
+	resultUniqueFields := make(map[string]string)
+
+	for _, fieldName := range uniqueFields {
+		if !requiredMap[fieldName] {
+			return nil, fmt.Errorf(
+				`%w: unique property needs to be a required property, "%s" is not in "required"`,
+				ErrInvalidUniqueProperty,
+				fieldName,
+			)
+		}
+
+		prop, ok := propertiesMap[fieldName]
+		if !ok {
+			return nil, fmt.Errorf(
+				`%w: missing property definition for unique field "%s"`,
+				ErrInvalidUniqueProperty,
+				fieldName,
+			)
+		}
+
+		fieldType, ok := prop.(map[string]interface{})["type"].(string)
+		if !ok || !isValidType(fieldType) {
+			return nil, fmt.Errorf(
+				`%w: invalid type "%s" for unique field "%s"`,
+				ErrInvalidUniqueProperty,
+				fieldType,
+				fieldName,
+			)
+		}
+
+		resultUniqueFields[fieldName] = fieldType
+	}
+
+	return &UniqueConstraintSchema{resultUniqueFields, uc.ERD, uc.ResourceID, uc.ctx, uc.db}, nil
+}
+
+// Checks if the provided field type is valid for unique constraints
+func isValidType(fieldType string) bool {
+	_, ok := map[string]bool{
+		"string": true, "number": true, "integer": true, "boolean": true,
+	}[fieldType]
+
+	return ok
```

Comment:
Is this any better than just returning from a conditional?

```suggestion
	return fieldType == "string" || fieldType == "number" || fieldType == "integer" || fieldType ==  "boolean"
```

Reply:
yes, I was originally thinking to use a set, then realize go doesn't have sets, ended up over thinking this